tsl/src/continuous_aggs/invalidation_multi.c

```diff
@@ -0,0 +1,503 @@
+	}
+
+	return caggs;
+}
+
+/*
+ * Fetch a hypertable entry for update
+ *
+ * This will fetch the hypertable entry for the given hypertable, and
+ * construct it if it was not there.
+ *
+ * Returns a pointer to the entry that was either found or added.
+ */
+MultiInvalidationEntry *
+multi_invalidation_state_hypertable_entry_get_for_update(MultiInvalidationState *state,
+														 int32 hypertable_id)
+{
+	bool hypertables_entry_found;
+	MultiInvalidationEntry *hypertables_entry =
+		hash_search(state->hypertables, &hypertable_id, HASH_ENTER, &hypertables_entry_found);
+
+	/* If hypertable entry was not found, initialize it. */
+	if (!hypertables_entry_found)
+	{
+		List *caggs = multi_invalidation_hypertable_entry_init(hypertables_entry, hypertable_id);
+
+		/* We know that the continuous aggregates for the hypertable were not
+		 * in the cache since the hypertable was not in the cache, so let's
+		 * write the associated continuous aggregates (that are all using the
+		 * WAL) to the ranges cache at this point. */
+		ListCell *lc;
+		foreach (lc, caggs)
+		{
+			ContinuousAgg *cagg = lfirst(lc);
+			int32 materialization_id = cagg->data.mat_hypertable_id;
+			bool ranges_entry_found;
+			MultiInvalidationRangeEntry *ranges_entry =
+				hash_search(state->ranges, &materialization_id, HASH_ENTER, &ranges_entry_found);
+
+			Assert(!ranges_entry_found);
+			multi_invalidation_range_entry_init(ranges_entry,
+												materialization_id,
+												cagg->bucket_function);
+		}
+	}
+
+	Assert(hypertables_entry != NULL);
+	return hypertables_entry;
+}
+
+void
+multi_invalidation_range_add(MultiInvalidationState *state, const Invalidation *invalidation)
+{
+	MultiInvalidationEntry *hypertables_entry =
+		multi_invalidation_state_hypertable_entry_get_for_update(state, invalidation->hyper_id);
+	TypeCacheEntry *typcache = lookup_type_cache(INT8RANGEOID, TYPECACHE_RANGE_INFO);
+
+	MemoryContext old_context = MemoryContextSwitchTo(state->context);
+
+	TS_DEBUG_LOG("CurrentMemoryContext: %s", CurrentMemoryContext->name);
+
+	/* Iterate over all associated caggs of the hypertable, expand the range
+	 * to the bucket boundaries for the continuous aggregate, and add it to
+	 * the ranges to build. */
+	ListCell *lc;
+	foreach (lc, hypertables_entry->caggs)
+	{
+		const int32 materialized_id = lfirst_int(lc);
+
+		/* We copy all the fields of the invalidation, but modify the hypertable id to be the id of
+		 * the materialization table. */
+		Invalidation cagg_invalidation = *invalidation;
+		cagg_invalidation.hyper_id = materialized_id;
+
+		bool ranges_entry_found;
+		MultiInvalidationRangeEntry *ranges_entry =
+			hash_search(state->ranges, &materialized_id, HASH_FIND, &ranges_entry_found);
+
+		Assert(ranges_entry->ranges_count < ranges_entry->ranges_alloc);
+
+		/* The continuous aggregate should be in the table since we added it when we added the
+		 * hypertable. If it is not there, something is wrong. */
+		Assert(ranges_entry_found);
```

Comment:
Does it make more sense to have this assert above the previous one since you want to ensure the entry is valid before it is used?

Reply:
Moved.